packages/web/src/containers/index.tsx

```diff
@@ -0,0 +1,33 @@
+import React from "react";
+import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
+import { Developer, Error, Home, Login, Projects } from "./pages";
+import { TradulabRouter, AuthMiddleware, RoleMiddleware } from "./routes";
+import { TradulabHeader } from "../components";
+
+const Routes: React.FC = () => (
+  <Router>
+    <Switch>
+      <Route component={Error} exact path="/error" />
+      <Route component={Login} exact path="/login" />
+      <TradulabRouter
+        Component={Developer}
+        Parent={TradulabHeader}
+        middlewares={[AuthMiddleware, RoleMiddleware]}
+        path="/dev"
+      />
+      <TradulabRouter
+        Component={Projects}
+        middlewares={[AuthMiddleware]}
+        path="/projects"
+      />
+      <TradulabRouter
+        Component={Home}
+        Parent={TradulabHeader}
+        middlewares={[AuthMiddleware]}
+        path="/"
+      />
```

Comment:
Achei mto complexo esse container pra pouca regra que ele introduz. O auth middleware pode ser só um container encapsulando todos os children autenticados.

```
<Header>
  <AuthMiddleware>
    // todas as rotas autenticadas aqui dentro
    <RoleMiddleware role="developer">
    // todas as rotas de um role aqui dentro
    </RoleMiddleware>
  </AuthMiddleware>>
</Header>
```

talvez precise usar mais o switch dessa maneira, mas eu acho mais fácil do que ter que implementar um contrato fixo de middleware.

Reply:
Calma, não entendi, você quer que mude toda a lógica que fiz de roteamento? Excluir o TradulabRouter? Eu não acho que ficou complexo, é uma recursao simples que é feita no TradulabRouter, fora que do jeito que fiz podemos passar em apenas um só parâmetros todas as lógicas necessárias para se executar ao entrar na rota, sem ter que ficar aninhado nada. Mas estou aberto a mudanças, se vc achar esta a melhor forma, posso mudar o roteamento.

Reply:
Neste caso, você preferia que AuthMiddleware fosse um container que renderizaria um Router? Fazer esse aninhamento de swich não vai ser complicado? Pois estariamos renderizando pedaços de ROUTER em lugares diferentes, pode atrapalhar no SWITCH na hora de achar as rotas, não?

Reply:
Do jeito que eu fiz, a única coisa que tem que ser feita é criar um novo middleware para cada nova regra de negócio que surgir entre as rotas. Ai é só ir no arquivo de rotas e renderizar um TradulaRouter passando todos os middlewares necessários. Ainda não vi nenhum ganho no que você pensou e não gostei dessa maneira de aninhar as rotas, talvez não entendi muito bem, acho que vamos ter que conversar sobre hehehe Fika pra próxima reunião <3

Reply:
Você entendeu o que o TradulabRouter faz? Ele executa todos os middlewares passados e a medida que a regra em cada middleware é antendida, ele renderiza outro TradulabRouter, só que agora com menos middlewares, no final, quando não houver mais middlewares ele, por fim, renderiza a rota.

Reply:
Além do mais, não entendo pq separar os middlewares desse jeito se podemos chama-los em um único parâmetro com uma função recursiva, pois no final, estes middlewares do seu jeito ou do meu, vão estar sendo chamados na mesma "frequencia", ou seja, sempre que a rota for chamada. Ou não?